### PR TITLE
Various BAO create/add cleanups to use writeRecord()

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Removes a bunch of boilerplate code from the BAOs.

Before
----------------------------------------
Boilerplate.

After
----------------------------------------
writeRecord()

### CODE REVIEW

File: CRM/Batch/BAO/Batch.php [hunk 1]
```diff
@@ -42,20 +42,10 @@ class CRM_Batch_BAO_Batch extends CRM_Batch_DAO_Batch {
    *   $batch batch object
    */
   public static function create(&$params) {
-    $op = 'edit';
-    $batchId = $params['id'] ?? NULL;
-    if (!$batchId) {
-      $op = 'create';
+    if (empty($params['id']) && empty($params['name'])) {
       $params['name'] = CRM_Utils_String::titleToVar($params['title']);
     }
```
Comment (colemanw) [inline CRM/Batch/BAO/Batch.php:47]: I've fixed this to only set a default if the param was not provided, assuming it to be an oversight.

File: CRM/Pledge/BAO/PledgeBlock.php [hunk 2]
```diff
@@ -79,49 +79,11 @@ public static function &create(&$params) {
...
-
-      $freqUnits = $newFreqUnits;
-      if (is_array($freqUnits) && !empty($freqUnits)) {
-        $freqUnits = implode(CRM_Core_DAO::VALUE_SEPARATOR, array_keys($freqUnits));
-        $pledgeBlock->pledge_frequency_unit = $freqUnits;
-      }
-      else {
-        $pledgeBlock->pledge_frequency_unit = '';
-      }
-    }
```
Comment (colemanw) [inline CRM/Pledge/BAO/PledgeBlock.php:112]: Giant code block replace with a one-liner; all those loops were doing boiled down to `array_keys(array_filter())`. The implode will be taken care of by `writeRecord()` per serialization metadata.

File: CRM/Pledge/BAO/PledgePayment.php [hunk 3]
```diff
@@ -154,31 +154,11 @@ public static function create($params) {
...
-    }
-
-    $payment = new CRM_Pledge_DAO_PledgePayment();
-    $payment->copyValues($params);
-
     // set currency for CRM-1496
-    if (!isset($payment->currency)) {
-      $payment->currency = CRM_Core_Config::singleton()->defaultCurrency;
+    if (empty($params['id']) && !isset($params['currency'])) {
+      $params['currency'] = CRM_Core_Config::singleton()->defaultCurrency;
```
Comment (colemanw) [inline CRM/Pledge/BAO/PledgePayment.php:159]: I think devs sometimes get confused that a function called "add" can also update. This is not the first time I've seen defaults get set even for update ops. I assume this was an oversight so I've fixed it by adding the `empty($params['id'])` condition.

**Comment by colemanw** [comment]
retest this please

**Comment by seamuslee001** [comment]
@colemanw test fails relate

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
retest this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
test this please

**Comment by colemanw** [comment]
test this please

**Pull request merged**